Lca adders

m-jamieson opened the pull request:
Added PhysicalQuantities module to support unit conversions
coal_upstream.py
Refined the EIA 923 data loading to support saving and reading from csvs
Refined EIA 7a downloading to download into appropriate sub-folders
Added/refined columns for incorporation into larger dataframe

m-jamieson committed:
renamed upstream integration to natural_gas_upstream.py
Added main routine to natural_gas_upstream
Cleaned up for column limit

Commit:
Adding routine to build up coal sources for each plant
using eia 923 and 7a data

m-jamieson committed:
Added a module to handle unit conversions, etc.

m-jamieson committed:
Modifications made to save/load data from csv to speed up processing
Modified/Added column names to support incorporation into total dataframe

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -2,8 +2,178 @@
+
+def eia_7a_download(year, save_path):
+    eia7a_base_url = 'http://www.eia.gov/coal/data/public/xls/'
+    name = 'coalpublic{}'.format(year)+'.xls'
```
FYI that you can don't need the string concat. `'coalpublic{}.xls'.format(year)` will work.

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -2,8 +2,178 @@
+    except:
+        print('Error downloading eia-7a: try manually downloading from:\n'+
+              url)
+def _clean_columns(df):
```
Might consider putting this function in util.py. I wasn't changing column names in any of Wes' code but it could be of use in the future.

m-jamieson replied:
Yeah - my plan was to leave that sort of stuff up to you for the time being. I have somewhat less visibility of the larger picture, and I'm not too sure what might be useful elsewhere.

Along the same lines, I'm not sure if adding a separate module for the eia7a download would be worthwhile.

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -2,8 +2,178 @@
+
+def _coal_code(row):
+    #_coal_code_str = basin_codes[basin] + '-' + coal_type_codes[coal_type] + '-' + mine_type
+    coal_code_str=(basin_codes[row['netl_basin']]+'-'+
```
Kudos for line wrapping! My suggested format would be:

```python
coal_code_str = (
    f'{basin_codes[row['netl_basin']]}-'
    f'{coal_type_codes[row['energy_source']]}-'
    f'{row['coalmine_type']}'
)
```

Change not required, up to you.

m-jamieson replied:
Done - had to use double quotes for the column names.

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -133,25 +307,32 @@ def generate_upstream_coal():
+            id_vars = ['plant_id','coal_source_code'], 
             var_name = 'FlowName', 
             value_name = 'FlowAmount')
+    melted_database_transport['coal_source_code']=melted_database_transport.apply(
```
Use `.map()` method, which performs a vector operation, rather than `.apply` with a function (has to loop through every row).

```python
# Tossing the dict back in here, can keep at top of module
transport_dict={'Avg Barge Ton*Miles':'Barge',
                'Avg Lake Vessel Ton*Miles':'Lake Vessel',
                'Avg Ocean Vessel Ton*Miles':'Ocean Vessel',
                'Avg Railroad Ton*Miles':'Railroad',
                'Avg Truck Ton*Miles':'Truck'}

melted_database_transport['coal_source_code'] = (
    melted_database_transport['coal_source_code'].map(transport_dict)
)
```

Looks like this is renaming the column rather than creating a new column. Is that correct?

m-jamieson replied:
Done.

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -2,8 +2,178 @@
+from os.path import join
+from electricitylci.utils import find_file_in_folder
+import requests
+import PhysicalQuantities as pq
```
Will need to add PhysicalQuantities as a requirement in setup.py. Have you considered adding conversion constants to globals.py instead? How often are you actually using pq?

m-jamieson replied:
Well, I think you've seen the extent to which pq is used so far. I just reworked someone else's fork the other day to get it to work with Python 3. I'm sure we could probably get away with 3-4 constants to handle most of the conversions we have. I just like to have non-rounded conversions present whenever I can. Anyhow, if PhysicalQuantities.py is present in the module, then no such need right? actually, I guess then it should be:

import electricitylci.PhysicalQuantities as pq.

I added in pq.py so that we could avoid having to add another package (such as pint).

gschivley replied:
Ok, that works.

gschivley commented on electricitylci/natural_gas_upstream.py:
```diff
@@ -0,0 +1,142 @@
+    #Import the mapping file which has the source gas basin for each Plant Id. 
+    #Merge with ng_generation dataframe.
+    
+    ng_basin_mapping = pd.read_csv(data_dir + 'gas_supply_basin_mapping.csv')
```
Is the gas_supply_basin_mapping.csv file included in this PR?

m-jamieson replied:
Not yet, I'm actually re-working the gas a bit now. Think I'm going to make the mapping county-based.

m-jamieson replied:
Done.

gschivley commented on electricitylci/natural_gas_upstream.py:
```diff
@@ -0,0 +1,142 @@
+            data_dir + 'NG_LCI.xlsx', 
+            sheetname = 'Basin_Mean_Data')
+    
+    ng_lci_air = ng_lci[ng_lci['Compartment'] == 'Air']
```
Use .loc here. Never know when you'll get screwed up with a view vs copy. And add the `, :` for all columns - I think it might return a view otherwise.

```python
ng_lci_air = ng_lci.loc[ng_lci['Compartment'] == 'Air', :]
```

m-jamieson replied:
This actually went away.

gschivley commented on electricitylci/natural_gas_upstream.py:
```diff
@@ -0,0 +1,142 @@
+    ng_lci_air = ng_lci[ng_lci['Compartment'] == 'Air']
+    
+    #Getting list of column values which are basically the emissions names
+    col_list = list(ng_lci.columns.get_values())
```
I'm pretty sure you don't need both of these lines.

```python
emissions_list = ng_lci.columns[3:].tolist()
```
should suffice.

m-jamieson replied:
Done.

gschivley commented on electricitylci/natural_gas_upstream.py:
```diff
@@ -0,0 +1,142 @@
+    emissions_list = [emission.strip(' (kg/MJ)') for emission in emissions_list]
+    
+    
+    stage_list = list(ng_lci['Stage'].unique())
```
`stage_list` doesn't appear to be used anywhere else.

gschivley commented on electricitylci/natural_gas_upstream.py:
```diff
@@ -0,0 +1,142 @@
+    ng_lci_basin = ng_lci_basin.drop(
+            ['NG_LCI_Name', 'Elec Fuel Consumption MMBtu'], axis = 1)
+    
+    #Output is kg emission for 2016 by facility Id, not normalized to 
```
Remove year from comment since it'll change.

m-jamieson replied:
Done.

gschivley reviewed (commented):
@jump2conclusionsmatt nice job getting this in so fast. Check out my comments. I didn't try to run any of the code - am assuming that you've made sure everything works.

m-jamieson committed:
General code clean up

Made the year an input rather than using the global variables

Kept the basin names in the output in case those are needed for further
analysis

m-jamieson committed:
Adding the upstream petroleum refining emissions

m-jamieson committed:
Added some columns to the fuel receipts to support use in other modules
specified PhysicalQuantities as a sub-module of electricitylci

jump2conclusionsmatt commented:
I pushed more commits to the lca_adders branch this morning. Supposedly, that is supposed to also add commits to this branch (Add more commits by pushing to the lca_adders branch on jump2conclusionsmatt/ElectricityLCI), but they're not showing up. Not sure if you have to initiate something for that.

gschivley commented on electricitylci/petroleum_upstream.py:
```diff
@@ -0,0 +1,108 @@
+    expected_lci_folder = 'petroleum_inventory'
+    for fuel in fuels:
+        for padd in padds:
+            fn = f'PRELIM_Mixer__{fuel}___PADD_{padd}_.xlsx'
```
Reminder that the data files will need to be included in this PR

gschivley commented on electricitylci/petroleum_upstream.py:
```diff
@@ -0,0 +1,108 @@
+    
+    Parameters
+    ----------
+    None
```
Add year parameter. It's pretty obvious what it's supposed to be but better to not have the conflicting info.

```python
year : int
    Year of EIA-923 fuel data to use
```

gschivley commented on electricitylci/coal_upstream.py:
```diff
@@ -83,17 +83,17 @@ def read_eia923_fuel_receipts(year):
-        csv_path = os.path.join(expected_923_folder,csv_file[0])
-        if os.path.exists(csv_path):
-            eia_fuel_receipts_df=pd.read_csv(csv_path)
+        if csv_file != []:
```
Can change this to just 
```python
if csv_file:
```
An empty list will return false, a list with objects will return true.

m-jamieson committed:
reformatted for column limit.

m-jamieson committed:
Adding --draft-- inventory files for upstream profiles.

m-jamieson committed:
Resolving review comments - minor changes.

m-jamieson committed:
Added a module to download cems data and aggregate data for the given year
This code is largely lifted from PUDL, however it's been modified to pull
daily data that is reported quarterly, resulting in a much smaller download

m-jamieson committed:
Added a module for generating upstream nuclear inventory. Note
that this simply applies a static upstream inventory to all plants.

m-jamieson committed:
Minor edits to clean up the petroleum output.

m-jamieson committed:
Made changes to inlcude the quantity of fuel used for later generating
unit processes from the overall inventory (i.e., emission per unit of coal)
Added code to the natural gas module to account for an updated natural
gas inventory.

m-jamieson committed:
Added some units and unit groups for the upstream unit processes

m-jamieson committed:
Added a module to turn the upstream data into dictionaries. The main
routine will also write the dictionaries to an olca json-ld zip.

m-jamieson committed:
Added a compartment for some of the VOCs

m-jamieson committed:
Renamed some columns

m-jamieson committed:
Added files to support upstream nuclear

m-jamieson committed:
Added a module to add geothermal life cycle inventory

m-jamieson committed:
Added support for geothermal

m-jamieson committed:
Added a last change date for the process writer

m-jamieson committed:
cleaning up comments

m-jamieson committed:
deleted unused submodules

m-jamieson committed:
Moved this to a csv

Commit:
Added a module that downloads state-level loss data and can also
generate a dataframe with aggregated loss rates by specified subregion

m-jamieson committed:
minor edits to ng upstream-changed fuel_type to GAS

m-jamieson committed:
Initial version of the solar pv inventory

m-jamieson committed:
initial version of the wind inventory

m-jamieson committed:
changed fuel_type to Oil

m-jamieson committed:
Initial version of electricity import impacts (Canada)

m-jamieson committed:
Initial version of module to combine plant and upstream data

m-jamieson committed:
data for canadian imports

m-jamieson committed:
Initial version of mapping NETL/openlca flows to fedelem

m-jamieson committed:
Duplicating index into BA Code column

Commit:
Code formatting, documentation. Added routine
to add fuel inputs to the generation database.

Commit:
Added returns to the docstrings

Commit:
Added the various routines for life cycle emissions

Commit:
Added lru_cache to speed up multiple calls to
eia923_download_extract.

Commit:
formatted with black. minor changes made
to support later routines

Commit:
Initial commit of alt_generation. This is an
attempt to provide an optimized aggregation
routine and separate routine to generate
plant-level emissions.

Commit:
Additions made to allow storage of generated
uuids and to return the modified dictionary.
Replaced the argument d with d_dict everywhere
because the use of d was interfering with
debugging.

Commit:
Changed to remove solar, wind, and geothermal
since those will be treated as power plants.

Commit:
Added a new example script to demonstrate
the new life cycle workflow.

Commit:
Formatted with black - I think that's it

Commit:
Changed to remove geothermal, solar, and wind

Commit:
Thought the uncommented version made
it into intial commit.

Commit:
Some required data files

m-jamieson committed:
Adjusted quantity to be in MJ

m-jamieson committed:
Changes made for performance

m-jamieson committed:
Some facilities were getting duplicated for some
reason leading to large coal inputs.

m-jamieson committed:
Auto naming of json-ld file so that I don't have
to remember to delete everytime.

m-jamieson committed:
Some fixes added for doing US generation.

m-jamieson committed:
I was getting errors trying to remove this
parameter from the config file

m-jamieson committed:
more error checking

m-jamieson committed:
fixed process naming

m-jamieson committed:
tried automating this a bit.

m-jamieson committed:
changing plant ids to ints

m-jamieson committed:
Changes made for new fedelem flows

m-jamieson committed:
Made a change to allow matching of
multiple patterns in utils.find_in_files.
This will allow storage of different
pages of eia923 data.

m-jamieson committed:
Chandes made for file pattern matching

m-jamieson committed:
New version of federal elementary flows
required changes to flow mapping

m-jamieson committed:
changes unit for natural gas to MJ

Commit:
Adding consumption mix to __init__.py

Commit:
Corrected BA and region name mapping

m-jamieson committed:
Started implementing logging.
Implemented Canadian generation for imports
Added some console messages

m-jamieson committed:
Implemented previous uncertainty calcs
Added logging
Added support for canadian generation
Added routines to aggregate multiple instances
of emissions species at a facility

m-jamieson committed:
fixed error in how duplicates were dropped

m-jamieson committed:
Made changes to support integration with
generation dataframe

m-jamieson committed:
Implemented some rounding - was running
into import errors in opnelca

Commit:
Revising code to import data from bulk_eia_data.py

Commit:
bulk_eia_data.py initial commit

m-jamieson committed:
Adding netl water use data

m-jamieson committed:
Added function to fill nans from other rows

Commit:
Converted code into function format

Commit:
Syntax changes to run combined build

Commit:
Working consumption mix funciton

Commit:
Combined build updates in progress

Commit:
Merge branch 'lca_adders' of https://github.com/jump2conclusionsmatt/ElectricityLCI into lca_adders

Commit:
Minor syntax changes

m-jamieson committed:
compartment cleanup. change to canadian generation

m-jamieson committed:
initial commit - hydro electric emissions

m-jamieson committed:
Changed units from ton-mi to kg-km

m-jamieson committed:
Implemented generation of T&D processes

m-jamieson committed:
Undid the rounding funtions

m-jamieson committed:
Small change for compartments

m-jamieson committed:
Data files for hydro power emissions

m-jamieson committed:
Removed data file from incorrect location

m-jamieson committed:
New file grabbed from gcooney

m-jamieson committed:
Formatted with black. Switched print for logging

Commit:
Minor update

Commit:
Consumption mix dictionary work

Commit:
071219 Pull from Jamieson

Commit:
initial commit

Commit:
Replaced some prints with logging

m-jamieson committed:
Various speed improvements and documentation

Commit:
Output directory change

Commit:
initial commit

Commit:
Merge branch 'lca_adders' of https://github.com/jump2conclusionsmatt/ElectricityLCI into lca_adders

m-jamieson committed:
Modifying the logging implementation

m-jamieson committed:
Made changes to support the ampd plants

m-jamieson committed:
Fixed path calculation

m-jamieson committed:
Changes made for speed

m-jamieson committed:
Formatted with black

m-jamieson committed:
Minor changes made to support ampd data

m-jamieson committed:
Hopefully these aren't used anymore

m-jamieson committed:
Made changes to read 860 and 923 data from files

Commit:
Corrected ferc grouping

Commit:
Changes made to ensure waste flows make it through

Commit:
Changed BA to reference Name instead of Code

Commit:
Removing this change I made to return total db

Commit:
Changes made so that this still works.

Commit:
Changing for new fedelemflows

Commit:
Reliability_Score to ReliabilityScore

Commit:
initial commit - plant construction

Commit:
Renamed file

Commit:
No longer used

Commit:
Added functions for consumption mix

Commit:
changes due to pandas version upgrade

Commit:
added function to write cons mix dict

Commit:
removed unnecessary code

Commit:
Added consumption mix function calls

m-jamieson committed:
Merge pull request #2 from gacooney/lca_adders

Lca adders

m-jamieson committed:
adding data files for ampd emissionsscript

m-jamieson committed:
Placed code under functions - multiple downloads
were happening before. Also commented out
a big chunk of code because it's repeated
in the bulk_eia_data module.

m-jamieson committed:
Only downloading bulk data if needed. Adding
BA names to the dataframe to match other areas
of the code.

m-jamieson committed:
initial commit - data files for trading module

m-jamieson committed:
Changed flow to match output of post t&d process

m-jamieson committed:
Added simple module for region selection and
added FERC region as option

m-jamieson committed:
Took out reading unused rows

m-jamieson committed:
moved output csvs to output_dir

m-jamieson committed:
Added documentation. Cleaned up __main__ routine.Added the ability to generate openlca dictionarieswithout the need for upstream dictionaries.

m-jamieson committed:
Formatted and added some documentation.

m-jamieson committed:
Changes to ensure that the old method for inventory
still works

m-jamieson committed:
Changed t&d to post consumption mix for consistency

m-jamieson committed:
Added a function to fill in UUIDs for input flows

m-jamieson committed:
Added a sample configuration for the new emissions

m-jamieson committed:
Added the routines to fill in UUIDs
However, not sure this is the ideal way.

m-jamieson committed:
Changes to account for other IDE files

m-jamieson committed:
Realized EIA region != EIA subregion

m-jamieson committed:
quick fixes based on some testing

gschivley commented on electricitylci/petroleum_upstream.py:
```diff
@@ -0,0 +1,118 @@
+            right=eia_fuel_receipts_df[['plant_id','heat_input','fuel_padd']],
+            left_on='fuel_code',
+            right_on='fuel_padd',
+            how='left').sort_values(['plant_id','fuel_padd','Flow.1'])
```
I'm getting an error here when running combinator.py. Some sort of keyerror for the column name Flow.1.
```
Traceback (most recent call last):
  File "combinator.py", line 379, in <module>
    petro_df = petro.generate_petroleum_upstream(2016)
  File "/Users/greg/Documents/NETL work/ElectricityLCI/electricitylci/petroleum_upstream.py", line 79, in generate_petroleum_upstream
    how='left').sort_values(['plant_id','fuel_padd','Flow.1'])
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/frame.py", line 4411, in sort_values
    stacklevel=stacklevel)
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/generic.py", line 1382, in _get_label_or_level_values
    raise KeyError(key)
KeyError: 'Flow.1'
```

m-jamieson replied:
This caused by a more recent version of pandas. When reading the Excel files with duplicate column names, pandas will append the .0, .1 etc. to the end of the column name to avoid that duplication - it does this even if you specify a range that doesn't include the duplicate column names. 

Just pushed a fix - added column rename to get rid of the suffixes.

m-jamieson committed:
Added column rename to avoid keyerrors on older
versions of pandas

m-jamieson committed:
Added a check to ensure a valid option is specified

gschivley commented on electricitylci/combinator.py:
```diff
@@ -0,0 +1,395 @@
+    )
+    plant_df = altg.create_generation_process_df()
+    plant_df["stage_code"] = "Power plant"
+    combined_df = concat_clean_upstream_and_plant(plant_df, upstream_df)
```
Getting another keyerror here, from the `fill_nans` function. Somehow the column `State` isn't included? 

```
INFO:ampd_plant_emissions.py:generate_plant_emissions:Summing eia923 fuel generation
INFO:ampd_plant_emissions.py:generate_plant_emissions:Summing eia923 boiler data
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating co2, ch4, n2o from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating so2 emissions from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating nox emissions from gen fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating co2, ch4, n2o emissions from boiler
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating so2 emissions from boiler fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Generating nox emissions from boiler fuel
INFO:ampd_plant_emissions.py:generate_plant_emissions:Choosing emission sources
INFO:ampd_plant_emissions.py:generate_plant_emissions:Melting and mapping
WARNING:combinator.py:concat_clean_upstream_and_plant:Error deleting column Net Generation (MWh)
Traceback (most recent call last):
  File "combinator.py", line 389, in <module>
    combined_df = concat_clean_upstream_and_plant(plant_df, upstream_df)
  File "combinator.py", line 265, in concat_clean_upstream_and_plant
    combined_df = fill_nans(combined_df)
  File "combinator.py", line 77, in fill_nans
    df[[key_column] + target_columns]
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/frame.py", line 2934, in __getitem__
    raise_missing=True)
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1354, in _convert_to_indexer
    return self._get_listlike_indexer(obj, axis, **kwargs)[1]
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1161, in _get_listlike_indexer
    raise_missing=raise_missing)
  File "/Users/greg/miniconda3/envs/elci/lib/python3.7/site-packages/pandas/core/indexing.py", line 1252, in _validate_read_indexer
    raise KeyError("{} not in index".format(not_found))
KeyError: "['State'] not in index"
```

m-jamieson replied:
Not at all sure how this happened - would guess that somehow, there ended up being a "State_x" or "State_y". Anyhow, added a check to ensure column is in the df to avoid the keyerror. If state is missing, it will make a new column and then use the existing code to populate nans.

Pushed fix.

m-jamieson committed:
Adding fix for missing columns - make "State"
column if necessary.
